### PR TITLE
Feature/enhance validation consistency for field elements

### PR DESCRIPTION
**Pull request by Sarin-Udompanish**
## Description

**From Field Elements API Specification changed:**

1. Introduce reportValidity(): A publicly accessible method that allows users to manually trigger validation. It returns a Boolean indicating whether the field is valid or not.
2. Introduce checkValidity(): Another publicly accessible method, similar to reportValidity(), which checks validity but doesn't display validation error messages. It also returns a Boolean indicating validity.

**Number Field Elements API Specification changed:**

_Validation on Input:_
The field element should also fire an "error-changed" event when the input value becomes valid after it was previously invalid. This ensures that users can respond to changes in validity in real-time.

**Other fields excepts Number Field Specifics changed:**

_Validation on Initialization:_
1. The field element should not perform validation automatically upon initialization. Users are responsible for initiating validation by calling reportValidity() or checkValidity() when needed.
2. minlength and maxlength validation should only occur when the field's value is changed by a user edit, not when set programmatically. This means that programmatically setting the value should not trigger these specific validation checks. ref: https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#setting-minimum-input-length-requirements:-the-minlength-attribute

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: documents/src/pages/elements/email-field.md [hunk 1]
```diff
@@ -90,9 +90,13 @@ emailField?.addEventListener("value-changed", (event) => {
 ```
 
 ## Input validation
-Validation occurs when the constraints are provided and the value changes. If the error state changes, it will dispatch the `error-changed` event along with the current error state.
+To validate input from users, `ef-email-field` provides similar features to a native input. When a user assigns an invalid input to the control, it will automatically apply an error style to alert the user. However, if you define a default value that is invalid, you need to call `input.reportValidity()` during initialization to ensure the error style is applied.
```
Comment (wsuwt) [inline documents/src/pages/elements/email-field.md]:
```suggestion
`ef-email-field` has validation logic similar to native input. When user types invalid value to the control, error style will be shown to notify user. However, if control is being initialised with an invalid value, `reportValidity()` must be called to ensure the error style is applied.
```

File: packages/core/__test__/elements/FormFieldElement.test.js [hunk 2]
```diff
@@ -213,6 +213,13 @@ describe('elements/FormFieldElement/EventsTest', function () {
     formFieldEl.inputElement.dispatchEvent(new CustomEvent('change'));
     expect(formFieldEl.changeEventCounter).to.equal(1);
   });
+  it('should fired error-changed event when call reportValidity with error stat with no validation', async function () {
```
Comment (wsuwt) [inline packages/core/__test__/elements/FormFieldElement.test.js]:
```suggestion
  it('should fired error-changed event when call reportValidity() when it is set as error with no constraint', async function () {
```

File: packages/elements/src/number-field/__test__/number-field.test.js [hunk 3]
```diff
@@ -575,6 +575,19 @@ describe('number-field/NumberField', function () {
       expect(el.error).to.equal(true);
       expect(el.value).to.equal('-20');
     });
+    it('Should fired error-changed event when value between valid / invalid', async function () {
```
Comment (wsuwt) [inline packages/elements/src/number-field/__test__/number-field.test.js]:
```suggestion
    it('Should fired error-changed event when error changes from invalid to valid', async function () {
```

File: packages/elements/src/text-field/__test__/text-field.validation.test.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
+import '@refinitiv-ui/elements/text-field';
+
+import '@refinitiv-ui/elemental-theme/light/ef-text-field';
+import { elementUpdated, expect, fixture, oneEvent } from '@refinitiv-ui/test-helpers';
+
+describe('text-field/Validation', function () {
+  describe('Default', function () {
+    it('should show error state on initial when set error attr', async function () {
```
Comment (wsuwt) [inline packages/elements/src/text-field/__test__/text-field.validation.test.js]:
```suggestion
    it('should show error state on initialisation when error attribute is set', async function () {
```

File: packages/elements/src/text-field/__test__/text-field.validation.test.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      const el = await fixture('<ef-text-field error></ef-text-field>');
+      expect(el.error).to.be.equal(true);
+    });
+    it('should remove error state when call reportValidity with no validation', async function () {
```
Comment (wsuwt) [inline packages/elements/src/text-field/__test__/text-field.validation.test.js]:
```suggestion
    it('should set error to false when reportValidity() is called with no validation constraint', async function () {
```

File: packages/elements/src/text-field/__test__/text-field.validation.test.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      expect(elMin.error).to.be.equal(false);
+      expect(elMax.error).to.be.equal(false);
+    });
+    it('should not show error state with invalid initial value', async function () {
+      const elMin = await fixture('<ef-text-field minlength="3" value="a"></ef-text-field>');
+      const elMax = await fixture('<ef-text-field maxlength="3 value="abcd"></ef-text-field>');
+      expect(elMin.error).to.be.equal(false);
+      expect(elMax.error).to.be.equal(false);
+    });
+    it('should show error state on initial when set error attr', async function () {
```
Comment (wsuwt) [inline packages/elements/src/text-field/__test__/text-field.validation.test.js]:
```suggestion
    it('should show error state on initialisation when error attribute is set', async function () {
```

File: packages/elements/src/text-field/index.ts [hunk 7]
```diff
@@ -137,6 +138,21 @@ export class TextField extends FormFieldElement {
     );
   }
 
+  /**
+   * Updates the element
+   * @param changedProperties Properties that has changed
+   * @returns {void}
+   */
+  protected override update(changedProperties: PropertyValues): void {
+    // This code probably should not be here, as validation must be instantiated by the app developer
```
Comment (wsuwt) [inline packages/elements/src/text-field/index.ts]: Is this comment still valid ?

**Review by wsuwt — CHANGES_REQUESTED** [review]

File: packages/core/__test__/elements/FormFieldElement.test.js (hunk 2 above)
Comment (Sarin-Udompanish) [reply packages/core/__test__/elements/FormFieldElement.test.js]: updated

File: documents/src/pages/elements/email-field.md (hunk 1 above)
Comment (Sarin-Udompanish) [reply documents/src/pages/elements/email-field.md]: updated

File: packages/elements/src/number-field/__test__/number-field.test.js (hunk 3 above)
Comment (Sarin-Udompanish) [reply packages/elements/src/number-field/__test__/number-field.test.js]: updated

File: packages/elements/src/text-field/__test__/text-field.validation.test.js (hunk 4 above)
Comment (Sarin-Udompanish) [reply packages/elements/src/text-field/__test__/text-field.validation.test.js]: updated

File: packages/elements/src/text-field/__test__/text-field.validation.test.js (hunk 5 above)
Comment (Sarin-Udompanish) [reply packages/elements/src/text-field/__test__/text-field.validation.test.js]: updated

File: packages/elements/src/text-field/__test__/text-field.validation.test.js (hunk 6 above)
Comment (Sarin-Udompanish) [reply packages/elements/src/text-field/__test__/text-field.validation.test.js]: updated

File: packages/elements/src/text-field/index.ts (hunk 7 above)
Comment (Sarin-Udompanish) [reply packages/elements/src/text-field/index.ts]: removed

**Review by wsuwt — APPROVED** [review]

**Comment by bualoy-napat** [comment]
LGTM!

**Pull request merged**